protocol/auditlog.go

```diff
@@ -0,0 +1,187 @@
+)
+
+type directoryHistory struct {
+	signKey   sign.PublicKey
+	snapshots map[uint64]*DirSTR
+	latestSTR *DirSTR
+}
+
+// A ConiksAuditLog maintains the histories
+// of all CONIKS directories known to a CONIKS auditor.
+// Each history includes the directory's public signing key
+// enabling the auditor to verify the corresponding signed
+// tree roots.
+type ConiksAuditLog struct {
+	histories map[string]*directoryHistory
+}
+
+func newDirectoryHistory(signKey sign.PublicKey, str *DirSTR) *directoryHistory {
+	h := new(directoryHistory)
+	h.signKey = signKey
+	h.snapshots = make(map[uint64]*DirSTR)
+	h.latestSTR = str
+	return h
+}
+
+// NewAuditLog constructs a new ConiksAuditLog. It creates an empty
+// log; the auditor will add an entry for each CONIKS directory
+// the first time it observes an STR for that directory.
+func NewAuditLog() *ConiksAuditLog {
+	l := new(ConiksAuditLog)
+	l.histories = make(map[string]*directoryHistory)
+	return l
+}
+
+// IsKnownDirectory checks to see if an entry for the directory
+// address addr exists in the audit log l. IsKnownDirectory() does not
+// validate the entries themselves. It returns true if an entry exists,
+// and false otherwise.
+func (l *ConiksAuditLog) IsKnownDirectory(addr string) bool {
+	h := l.histories[addr]
+	if h != nil {
+		return true
+	}
+	return false
+}
+
+// Insert creates a new directory history for the key directory addr
+// and inserts it into the audit log l.
+// The directory history is initialized with the key directory's
+// signing key signKey, a list of STRs representing the
+// directory's prior history oldSTRs, and the directory's latest STR
+// latestSTR.
+// Insert() returns an ErrAuditLog if the auditor attempts to create
+// a new history for a known directory, an ErrMalformedDirectoryMessage
+// if oldSTRs is malformed, and nil otherwise.
+// Insert() only creates the initial entry in the log for addr. Use Update()
+// to insert newly observed STRs for addr in subsequent epochs.
+// FIXME: pass Response message as param
+// masomel: will probably want to write a more generic function
+// for "catching up" on a history in case an auditor misses epochs
+func (l *ConiksAuditLog) Insert(addr string, signKey sign.PublicKey,
+	oldSTRs map[uint64]*DirSTR, latestSTR *DirSTR) error {
+
+	// error if we want to create a new entry for an addr we already know
+	if l.IsKnownDirectory(addr) {
+		return ErrAuditLog
+	}
+
+	// make sure we have oldEpochs unless we're inserting right
+	// at the start of a directory's history
+	if oldSTRs == nil && latestSTR.Epoch > 0 {
+		return ErrMalformedDirectoryMessage
+	}
+
+	// create the new directory history
+	h := newDirectoryHistory(signKey, latestSTR)
+
+	startEp := uint64(0)
+	endEp := latestSTR.Epoch
+
+	// add each old STR into the history
+	// This loop automatically catches if oldSTRs is malformed
+	// (i.e. oldSTRs starts at epoch 0 and
+	// len(oldSTRs) != latestSTR.Epoch-1)
+	for ep := startEp; ep < endEp; ep++ {
```

Comment:
There sort of already is one on L107. But that FIXME is a bit buried at the bottom of the function.